### PR TITLE
Add example states files and several py files

### PR DESCRIPTION
**Pull request by ryg-**
Signed-off-by: Roman Grigoryev <roman.grigorev@suse.de>

### CODE REVIEW

File: lib/flask_tasks.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+    """
+    @wraps(f)
+    def wrapped(*args, **kwargs):
+        def task(app, environ):
+            # Create a request context similar to that of the original request
+            # so that the task can have access to flask.g, flask.request, etc.
+            with app.request_context(environ):
+                try:
+                    # Run the route function and record the response
+                    logging.debug("____________________________")
```
Comment (kshtsk) [inline lib/flask_tasks.py]: Maybe you better create a dedicated logging method for drawing those lines automatically.

Comment (ryg-) [reply lib/flask_tasks.py]: Just deleted it

File: bin/pelagos.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+# process command line parameters
+def print_help():
+    print('pelagos.py -c <config file>')
+
+
+if __name__ == '__main__':
+    try:
+
+        opts, args = getopt.getopt(sys.argv[1:],
```
Comment (kshtsk) [inline bin/pelagos.py]: nit: maybe it is better to use argparse or docopt?

File: lib/flask_tasks.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+                    logging.debug("Executing target function")
+                    tasks[id]['started'] = True
+                    tasks[id]['rv'] = f(*args, **kwargs)
+                    tasks[id]['status'] = 'done'
+                    logging.debug("target function completed")
+                    logging.debug("____________________________")
```
Comment (kshtsk) [inline lib/flask_tasks.py]: same here

File: lib/flask_tasks.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+                    tasks[id]['rv'] = current_app.handle_http_exception(e)
+                    traceback.print_exc(file=sys.stdout)
+                except Exception as e:
+                    logging.debug("Caught exception:" + str(e))
+                    # The function raised an exception, so we set a 500 error
+                    tasks[id]['rv'] = InternalServerError()
+                    traceback.print_exc(file=sys.stdout)
+
+            #                    if current_app.debug:
+#                        # We want to find out if something happened so reraise
```
Comment (kshtsk) [inline lib/flask_tasks.py]: cleanup?

File: lib/flask_tasks.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+#                        raise
+                finally:
+                    # We record the time of the response, to help in garbage
+                    # collecting old tasks
+                    tasks[id]['endtime'] = timestamp()
+
+        # Assign an id to the asynchronous task
+        id = uuid.uuid4().hex
```
Comment (kshtsk) [inline lib/flask_tasks.py]: this looks like ambiguous usage of id() python function, maybe it is better to use 'task_id' variable name?

File: lib/flask_tasks.py [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+
+        # Record the task, and then launch it
+        #tasks[id]['started'] = False
+        tasks[id] = {'task': threading.Thread(
+            target=task, args=(current_app._get_current_object(),
+                               request.environ))}
+        tasks[id]['task'].start()
```
Comment (kshtsk) [inline lib/flask_tasks.py]: I would not use a dictionary lookup on each line for modifying an object, it is enough to capture the object once, and modify it's elements, check it out:
```
>>> a={'1': {'a': 'x'}, '2': {'a': 'y'}}
>>> a
{'1': {'a': 'x'}, '2': {'a': 'y'}}
>>> t=a['1']
>>> t['a']='s'
>>> t
{'a': 's'}
>>> a
{'1': {'a': 's'}, '2': {'a': 'y'}}
>>> 

```

File: lib/flask_tasks.py [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+        tasks[id]['starttime'] = timestamp()
```
Comment (kshtsk) [inline lib/flask_tasks.py]: Not sure, but this must go before actual `start()` invocation.

File: lib/flask_tasks.py (hunk 3 above)
Comment (ryg-) [reply lib/flask_tasks.py]: removed

File: lib/flask_tasks.py (hunk 4 above)
Comment (ryg-) [reply lib/flask_tasks.py]: done

File: lib/flask_tasks.py (hunk 5 above)
Comment (ryg-) [reply lib/flask_tasks.py]: done

File: lib/flask_tasks.py (hunk 7 above)
Comment (ryg-) [reply lib/flask_tasks.py]: done

File: lib/flask_tasks.py (hunk 6 above)
Comment (ryg-) [reply lib/flask_tasks.py]: will fix in future

File: lib/network_manager.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
+import json
+import logging
+
+logging.basicConfig(format='%(asctime)s | %(name)s | %(message)s',
+                    level=logging.DEBUG)
+
+
+data_file = '../networks_info/nue_ses_network.json'
```
Comment (kshtsk) [inline lib/network_manager.py]: What is this related to? Current file or directory?

Comment (ryg-) [reply lib/network_manager.py]: predefined path removed

File: test_pelagos_teuthology/test_pelagos.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+test_config = dict(pelagos=dict(
+    endpoint='http://localhost:5000/',
+    machine_types='type1,type2',
+))
+
+test_pxelinux_cfg_root_dir = '/tmp/tftp'
+test_pxelinux_cfg_dir = test_pxelinux_cfg_root_dir + '/pxelinux.cfg'
+
+class TestPelagios(object):
```
[the remark this answers is not in the document]
Comment (ryg-) [reply test_pelagos_teuthology/test_pelagos.py]: fixed

File: lib/hw_node.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+# current output after boot
+# -----------------------
+# Welcome to openSUSE Leap 15.0 - Kernel 4.12.14-lp150.11-default (ttyS1).
+#
+#
+# localhost login
+def wait_node_is_ready(node, timeout=5, attempts=120):
+    local = LocalNode()
+    local.hostname = 'local'
```
Comment (kshtsk) [inline lib/hw_node.py]: why not providing this to LocalNode constructor as argument?

Comment (ryg-) [reply lib/hw_node.py]: done

File: bin/make_cfgs.py [hunk 11]
```diff
@@ -40,7 +40,7 @@
         formatter_class=argparse.RawDescriptionHelpFormatter)
 
 
-def reorder_ip(ip):
+def reverse_ip(ip):
```
Comment (kshtsk) [inline bin/make_cfgs.py]: Because the ip is supposed to be always a string represented IPv4, consider to use this single line instead of regex:
```
   return '.'.join(reversed(ip.split('.')))
```

Comment (kshtsk) [inline bin/make_cfgs.py]: typo *reversed_ip

Comment (ryg-) [reply bin/make_cfgs.py]: done

Comment (ryg-) [reply bin/make_cfgs.py]: done

File: bin/pelagos.py (hunk 2 above)
Comment (ryg-) [reply bin/pelagos.py]: done

File: bin/make_cfgs.py [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+                "'" + n['comment'] + "'",
+            ]
+        nodes_sql_lines.append(
+            sql_line_prefix +\
+            " ( " + " , ".join(sql_update_body) + " ) ;\n"
+            #+\
+            #" WHERE name     = '" + n['node'] + "' ;\n"
+        )
+
+# ----------- pxe ---------------
```
Comment (kshtsk) [inline bin/make_cfgs.py]: Why this comment needed here?

File: bin/make_cfgs.py [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+with open(target_dir_prefix + pxe_output_file, 'w') as ofile:
+    ofile.write("\n".join(pxe_node_lines)+"\n")    
+    ofile.write("\n".join(pxe_node_ptr_lines)+"\n")
+print("File [{}] written".format(target_dir_prefix +
+                                    pxe_output_file))
+
+# ----------- conman ---------------
```
Comment (kshtsk) [inline bin/make_cfgs.py]: Why this comment needed here?

File: bin/make_cfgs.py [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+with open(target_dir_prefix + conman_tmpl, "r") as cf:
+    for line in cf:
+        line = line.replace("@@CONSOLES@@", consoles)
+        lines = lines + line
+
+with open(target_dir_prefix + conman_cfg, 'w') as ofile:
+    ofile.write(lines)
+print("File [{}] written".format(target_dir_prefix + conman_cfg))
+
+# ----------- roster ---------------
```
Comment (kshtsk) [inline bin/make_cfgs.py]: Why this comment needed here?

File: bin/make_cfgs.py [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+
+with open(target_dir_prefix + roster_file, 'w') as ofile:
+    ofile.write(roster_recors)
+print("File [{}] written".format(target_dir_prefix + roster_file))
+
+# ----------- sql file ---------------
```
Comment (kshtsk) [inline bin/make_cfgs.py]: Why this comment needed here?

File: bin/make_cfgs.py [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+                                      ))
+
+    if ('hsm_ip_type' in n.keys()) and (n['hsm_ip_type'] == 'dynamic'):
+        pxe_node_lines.append("dhcp-host={},{}-hsm,{}"
+                          .format(n['hsm_mac'], hn(n['node']), n['hsm_ip']))
+        pxe_node_ptr_lines.append("ptr-record={}.in-addr.arpa,{}-hsm"
+                              .format(reorder_ip(n['hsm_ip']), hn(n['node'])))
+
+    consoles = consoles +\
+               "\nCONSOLE name=\"{}\" IPMIOPTS=\"U:{},P:{}\" dev=\"ipmi:{}\""\
```
Comment (kshtsk) [inline bin/make_cfgs.py]: it is better to use named placeholder?

File: bin/make_cfgs.py [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+                   .format(hn(n['node']), bmc_user, bmc_pass, n['bmc_ip'])
+    roster_recors = roster_recors +\
+                    n['node']+":\n" +\
+                    "  host: " + n['ip'] + "\n" +\
+                    "  user: root\n"
```
Comment (kshtsk) [inline bin/make_cfgs.py]: Why don't you use here `.format`?

File: bin/make_cfgs.py [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+    if 't_exclude' in n.keys() and n['t_exclude'] != 'yes':
+        sql_update_body = [
+                "'" + n['node'] + "'",
+                "'" + n['t_machine_type'] + "'",
+                "'" + n['mac'] + "'",
+                "'f'",
+                "'f'",
+                "'x86-64'",
+                "'" + n['comment'] + "'",
+            ]
```
Comment (kshtsk) [inline bin/make_cfgs.py]: Exactly here is the best place to use `.format`

File: bin/make_cfgs.py [hunk 19]
```diff
@@ -0,0 +1,182 @@
...
+    if n['ip_type'] == 'dynamic':
+        pxe_node_lines.append("dhcp-host={},{},{}"
+                          .format(n['mac'], hn(n['node']), n['ip']))
+        pxe_node_ptr_lines.append("ptr-record={}.in-addr.arpa,{}.{}"
+                              .format(reorder_ip(n['ip']),
+                                      hn(n['node']),
+                                      domain))
+
+    if n['bmc_ip_type'] == 'dynamic':
+        pxe_node_lines.append("dhcp-host={},{}-bmc,{}"
```
Comment (kshtsk) [inline bin/make_cfgs.py]: Could you please use named place holder?

File: bin/make_cfgs.py [hunk 20]
```diff
@@ -0,0 +1,182 @@
...
+nodes_sql_lines = []
+
+for n in nodes:
+    print('Process node' + n['node'])
+    
+    if n['ip_type'] == 'dynamic':
+        pxe_node_lines.append("dhcp-host={},{},{}"
+                          .format(n['mac'], hn(n['node']), n['ip']))
+        pxe_node_ptr_lines.append("ptr-record={}.in-addr.arpa,{}.{}"
+                              .format(reorder_ip(n['ip']),
```
Comment (kshtsk) [inline bin/make_cfgs.py]: I thought it should be `reversed_ip` here.

File: bin/make_cfgs.py [hunk 21]
```diff
@@ -0,0 +1,182 @@
...
+    INSERT INTO nodes (name, machine_type, mac_address, is_vm, locked, arch, description)
+        VALUES  ( 'ses-client-3.a.b.de' , 'client' ,
+            'aa:bb:cc:dd:67:16' , 'f' , 'f' , 'x86-64' , '' ) ;
+"""
+
+parser = argparse.ArgumentParser(description=description,
+        formatter_class=argparse.RawDescriptionHelpFormatter)
+
+
+def reorder_ip(ip):
```
Comment (kshtsk) [inline bin/make_cfgs.py]: use reverse?

File: bin/make_cfgs.py [hunk 22]
```diff
@@ -0,0 +1,182 @@
...
+    ip_split = re.search('(\d+)\.(\d+)\.(\d+)\.(\d+)', ip)
+    print("ip_split:")
+    print(ip_split.groups())
+    return ip_split.groups()[3] + '.'\
+           + ip_split.groups()[2] + '.'\
+           + ip_split.groups()[1] + '.'\
+           + ip_split.groups()[0]
+
+# get host name from probably fqdn
+def hn(name):
```
Comment (kshtsk) [inline bin/make_cfgs.py]: hn=>short_hostname?
name=>fqdn?

File: bin/make_cfgs.py (hunk 12 above)
Comment (ryg-) [reply bin/make_cfgs.py]: removed

File: bin/make_cfgs.py (hunk 13 above)
Comment (ryg-) [reply bin/make_cfgs.py]: removed

File: bin/make_cfgs.py (hunk 14 above)
Comment (ryg-) [reply bin/make_cfgs.py]: removed

File: bin/make_cfgs.py (hunk 15 above)
Comment (ryg-) [reply bin/make_cfgs.py]: removed

File: bin/make_cfgs.py (hunk 21 above)
Comment (ryg-) [reply bin/make_cfgs.py]: and it was renamed, you could see in branch

File: bin/make_cfgs.py (hunk 22 above)
Comment (ryg-) [reply bin/make_cfgs.py]: done

File: bin/make_cfgs.py (hunk 20 above)
Comment (ryg-) [reply bin/make_cfgs.py]: it was renamed previously

File: bin/make_cfgs.py (hunk 19 above)
Comment (ryg-) [reply bin/make_cfgs.py]: I suppose 3 parameters in one string don't need placeholder

File: bin/make_cfgs.py (hunk 18 above)
Comment (ryg-) [reply bin/make_cfgs.py]: I don't see how '.format' could help.

File: bin/make_cfgs.py [hunk 23]
```diff
@@ -0,0 +1,182 @@
...
+pxe_node_ptr_lines = []
+nodes_sql_lines = []
+
+for n in nodes:
+    print('Process node' + n['node'])
+    
+    if n['ip_type'] == 'dynamic':
+        pxe_node_lines.append("dhcp-host={},{},{}"
+                          .format(n['mac'], hn(n['node']), n['ip']))
+        pxe_node_ptr_lines.append("ptr-record={}.in-addr.arpa,{}.{}"
```
[the remark this answers is not in the document]
Comment (ryg-) [reply bin/make_cfgs.py]: done

File: bin/make_cfgs.py (hunk 18 above)
Comment (ryg-) [reply bin/make_cfgs.py]: replaced to %s

File: bin/make_cfgs.py (hunk 17 above)
Comment (ryg-) [reply bin/make_cfgs.py]: done

File: bin/make_cfgs.py (hunk 16 above)
Comment (ryg-) [reply bin/make_cfgs.py]: done

**Comment by ryg-** [comment]
Merge
